### PR TITLE
Documentation: Added getting started documentation for ifcopenshell core

### PR DESCRIPTION
**Pull request by jonganej**
As suggested in #4351, here is an initial version of documentation for the core module of IfcOpenShell.

Please do let me know if there are things I should add or how I could improve this.

### CODE REVIEW

File: src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst
Comment (aothms) [inline src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst:46]: Do you think we should rewrite this using the Boost preprocessor mechanism (see below). Either way is a bit unfortunate, you either have a a bunch of if statements, or a incomprehensible bunch of preprocessor magic. I do think that the 2nd approach might in the end result in a better developer experience.

I think this would be the time to decide, as we're setting kind of the basis for the C++ docs here.

https://github.com/IfcOpenShell/IfcOpenShell/blob/9485190604bb3fed6f45199684609667bc2ed1df/src/examples/IfcParseExamples.cpp#L217-L227

File: src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst
Comment (aothms) [inline src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst:62]: Unfortunately this is still not consistent in the C++ side.

```suggestion
Note that optional properties like name, long name, or description, among others, are wrapped with a ``boost::optional`` (`documentation <https://www.boost.org/doc/libs/1_84_0/libs/optional/doc/html/index.html>`__) container object. Except for pointer types such as IfcRoot::OwnerHistory, here the developer is still responsible for performing comparison to `nullptr` before using the value.
```

File: src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst
Comment (aothms) [inline src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst:86]: I would use `x->template as<typename Schema::IfcRelDefinesByProperties>()` here.

Not sure if `template as` is required or `as` is sufficient. We need to check with gcc/clang.

Comment (aothms) [reply src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst:86]: Same for below. Since we're setting the basis here, better do as<...> instead of dynamic_cast<...>. Functionally it's the same, but we might implement it differently later on.

File: src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst
Comment (aothms) [inline src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst:109]:
```suggestion
The existence of certain fields and properties should also be checked.  Also note that IfcOpenShell has as of now not been tested explicitly against malicious inputs. Schema validation (the correctness of attribute types and conformance to schema where rules) can currently only be assessed in Python (using `ifcopenshell.validate --rules`).
```

**Review by aothms — CHANGES_REQUESTED** [review]
Super awesome. Thanks for taking the time to write this down. See my comments inline.

File: src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst
Comment (jonganej) [reply src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst]: `template as` is necessary. `as` doesn't seem to compile

File: src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst
Comment (jonganej) [reply src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst:46]: Hmm, I've never used these Boost preprocessor macros before. I see the advantage of macros, but for both cases, one still has to explicitly enumerate all the available schemas either with a `#define SCHEMA_SEQ` or a `else if `. In my production code, I chain `if`, `else if` and end with a `else` to catch the case where the input file does not match any of those schemas. Would that be possible too with the preprocessor macros?

Comment (jonganej) [reply src/ifcopenshell-python/docs/ifcopenshell/getting_started.rst]: Ok, I'm using preprocessor macros now in an `if-else if` fashion with a final `else` to ensure that one can fail explicitly if the schema is not recognized. We still have the enumerated `#define` for the schemas, but at least it's clearer now. I tested this approach on our productive use-case and it works.

**Comment by Moult** [comment]
Good to merge?

**Comment by jonganej** [comment]
I would say so. I’ve done the necessary changes @aothms suggested. Was wondering if he or one of the other maintainers wanted to take a look at it.

**Comment by Moult** [comment]
I'll merge it and if there are more improvements it can be fixed over time :) Thank you so much!

**Comment by aothms** [comment]
Thanks again. Sorry, this got buried in my pile of holiday emails.

**Pull request merged**
